src/main/java/com/fibermc/essentialcommands/types/WarpStorage.java

```diff
@@ -22,27 +27,60 @@ public WarpStorage(NbtCompound nbt) {
         loadNbt(nbt);
     }
 
-    @Override
+    public static WarpStorage fromNbt(NbtCompound nbt) {
+        // Try codec first
+        var result = CODEC.parse(NbtOps.INSTANCE, nbt);
+        if (result.isSuccess()) {
+            return result.getOrThrow();
+        }
+
+        // Fallback to legacy parsing
+        WarpStorage storage = new WarpStorage();
+        storage.loadNbt(nbt);
+        return storage;
+    }
+
     public NbtCompound writeNbt(NbtCompound nbt) {
-        this.forEach((key, value) -> nbt.put(key, value.asNbt()));
-        return nbt;
+        var result = CODEC.encode(this, NbtOps.INSTANCE, nbt);
+
+        if (result.isSuccess()) {
+            return result.getOrThrow().asCompound().orElseThrow();
+        }
+
+        throw new RuntimeException("Failed to encode WarpStorage to NBT: " + result.error());
     }
 
     /**
      * @param nbt NbtCompound or NbtList. (Latter is deprecated)
      */
-    public void loadNbt(NbtElement nbt) {
+    private void loadNbt(NbtElement nbt) {
         if (nbt.getType() == 9) {
             // Legacy format
             NbtList homesNbtList = (NbtList) nbt;
             for (NbtElement t : homesNbtList) {
                 NbtCompound homeTag = (NbtCompound) t;
                 String name = homeTag.getString("homeName").orElseThrow();
-                super.put(name, WarpLocation.fromNbt(homeTag, name));
+                var location = MinecraftLocation.fromNbt(homeTag);
+                super.put(
+                    name,
```

Comment:
technically, the old and new Compound varieties should be format-compatible, as the new one is fairly forgiving. The NbtList one is can stick around for compatibility until if/when Mojang decides to break the methods used here in a complicated way.